### ISSUE TITLE
aws s3 sync traverses excluded directory - takes up a long time

### ISSUE THREAD

**Issue by ChrisSLT** [aws/aws-cli#1138]
This has been discussed elsewhere but I wasn't sure what the response was.  I wanted to see if there would be an ability to stop SYNC from traversing an excluded directory because it takes up a lot of time.

When you do a sync from ~/ with --exclude "STAR" and --include "~/Documents/STAR" for example, the sync operation still takes many minutes to complete because it walks through every single file in every directory in the user's home folder.  I wish I knew enough to get into the actual code, but I'm wondering if you could stop the SYNC operation from doing this?

Thank you

**Comment by jamesls** [comment]
I agree, we should stop traversing into a directory if it's excluded/included.  This is also related to #1117 which will throw warnings even if a file is excluded, which has the same root cause:  we're traversing into a directory when we could be skipping it.

I don't see any reason why we shouldn't change this behavior.

**Comment by ChrisSLT** [comment]
Terrific, thank you

**Comment by ChrisSLT** [comment]
I know you guys are working away on many fronts, but wondering if you could let us know a timeline or roadmap on fixing this issue?  thank you much

**Comment by ChrisSLT** [comment]
Would you be able to let us know a timeline on this?

**Comment by kyleknap** [comment]
It is difficult to put a specific timeline/date for this. Since this would fix a couple of bugs that have been reported to us, it is has a relatively high priority for us. However, implementation-wise it may take a while because it requires a good amount of refactoring on the internals of the s3 commands. Ultimately, the feature is near the top of our list of features to implement, but time has yet to be allocated to work on it. We can let you know when we do start working on this feature.

**Comment by ChrisSLT** [comment]
Thank you.  Really appreciate it.

**Comment by ChrisSLT** [comment]
Can you please give an indication as to when this issue will be fixed?  It creates a lot of complexity to have to have multiple scripts to get around this bug.  Thank you

**Comment by ChrisSLT** [comment]
Can you tell me if this is intended to be fixed?

**Comment by ChrisSLT** [comment]
Please update us on this issue.  Due to the bug, our team is getting pressure to move away from S3 sync.

**Comment by ASayre** [comment]
Good Morning!

We're closing this issue here on GitHub, as part of our migration to [UserVoice](https://aws.uservoice.com/forums/598381-aws-command-line-interface) for feature requests involving the AWS CLI.

This will let us get the most important features to you, by making it easier to search for and show support for the features you care the most about, without diluting the conversation with bug reports.

As a quick UserVoice primer (if not already familiar): after an idea is posted, people can vote on the ideas, and the product team will be responding directly to the most popular suggestions.

We’ve imported existing feature requests from GitHub - Search for this issue there!

And don't worry, this issue will still exist on GitHub for posterity's sake.  As it’s a text-only import of the original post into UserVoice, we’ll still be keeping in mind the comments and discussion that already exist here on the GitHub issue.

GitHub will remain the channel for reporting bugs.  

Once again, this issue can now be found by searching for the title on: https://aws.uservoice.com/forums/598381-aws-command-line-interface 

-The AWS SDKs & Tools Team

**Comment by jamesls** [comment]
Based on community feedback, we have decided to return feature requests to GitHub issues.

**Comment by minjcho** [comment]
Opened #10273 which hopefully resolves this issue

Approach: a sound `Filter.can_skip_directory(dir_path)` that prunes a subtree only when the filter chain *provably* cannot include any descendant. The PR #5425 regression case (`--exclude '*' --include '*.py'` silently dropping nested `.py` files, [discussion_r883991435](https://github.com/aws/aws-cli/pull/5425#discussion_r883991435)) is handled automatically — the include pattern is detected as possibly matching descendants of any subtree, blocking the prune.

A 1M-file directory excluded via `--exclude 'src/*'` drops from ~70s to ~50ms. Existing match logic in `filters.py` is untouched.

Thanks to @clarete and @bugfood whose prior attempts (#2105, #5425) clarified the design constraints.
